### DISCUSSION TITLE
Request: ui.tutorial()

### DISCUSSION THREAD

**Discussion by HilkopterBob** [zauberzeug/nicegui#680] [Ideas / Feature Requests]
Hey, im currently building a quite complex wrapper for ansible (network automation platform) for my company.
I tryed to make everything as simple as possible and i wrote a quite long documentation, but i know that most end-users would love to have an interactive tutorial!

as a first concept i would suggerst the following:
programmer pov:
you can mark certain ellements with a `tutorial_id`, for example: 
```python
ui.button("Hello There!", on_click=greet_user, tutorial_id=1)
``` 

to display information about a tutorial marked element you have to define elements that shuld be displayed:
```python
with tutorial_view(id=1, interactive=True, reveal_element=True):
    ui.label("If you click that button, it will greet you!")
```

if you marked all elements you want with a `tutorial_id` and created all `tutorial_views()`
you can start the tutorial with `ui.tutorial()`

user pov:

if the tutorial is started the sceen gets overlayed and interactions are blocked, in (mostly) the same way as `ui.dialog` does. In fact, the `tutorial_view`'s them selves are dialogs with their defined content. In the example tutorial i made here, the user gets  the dialog, wich explanes what the "Hello there" button does. Because of the `interactive=True` flag, the button does not get overlayed gray and is still clickabel and executes normaly on `on_click`. The button also gets highlighted by a border (because `reveal_element=True`) in the primary or secondary (or custom) color, so its more visible to the user.

---
overall its a rough scetch, but i think a simple way to make a (long) tutorial would be a great addition, not only for the user, but for the developer(s), to lift simple explainatory work from their shoulders.

**Comment by falkoschindler** [comment]
Interesting idea, @HilkopterBob!

So you basically pick UI elements that should be part of the tutorial, define UI elements that should appear to explain these elements, and finally you somehow start the tutorial which cycles through all its elements.

That's probably something one could implement without the need for changing the NiceGUI API. Once it works well, we can include it in NiceGUI.

In order to keep the tutorial feature relatively isolated from the core implementation, I would start with something like this:
```py
start = ui.button('Start')
stop = ui.button('Stop')

tutorial = ui.tutorial()
with tutorial.step(start):
    ui.label('This button starts the calculation.')
with tutorial.step(stop):
    ui.label('This button stops the calculation.')

ui.button('Start tutorial', on_click=tutorial.start)
```

The tutorial steps could, of course, get more parameters like `interactive` or `reveal_element`. But I would start small. 😉

I am, however, not sure how the visualization of a step might work. Where would you show the labels? Maybe you can draft

**Reply by HilkopterBob** [reply]
Hey im not good in coding, so i wont be able to even scetch a working implementation for you, i am sorry for that. If you want i can try to edit some screenshots so they visualize my initial idea

**Reply by falkoschindler** [reply]
No worries. But yes, some graphical explanation would certainly help to explain where these labels would end up.
Maybe we don't even need labels (and other elements) per step, but tooltips would already be enough?
